rust/kernel/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! The `kernel` crate
+
+#![no_std]
+#![feature(allocator_api, alloc_error_handler, const_raw_ptr_deref)]
+
+extern crate alloc;
+
+use core::panic::PanicInfo;
+
+mod allocator;
+pub mod bindings;
+pub mod c_types;
+pub mod chrdev;
+mod error;
+pub mod file_operations;
+pub mod filesystem;
+pub mod prelude;
+pub mod printk;
+pub mod random;
+pub mod sysctl;
+mod types;
+pub mod user_ptr;
+
+pub use crate::error::{Error, KernelResult};
+pub use crate::types::{CStr, Mode};
+
+/// Declares the entrypoint for a kernel module. The first argument should be a type which
+/// implements the [`KernelModule`] trait. Also accepts various forms of kernel metadata.
+///
+/// Example:
+/// ```rust,no_run
+/// use kernel::prelude::*;
+///
+/// struct MyKernelModule;
+/// impl KernelModule for MyKernelModule {
+///     fn init() -> KernelResult<Self> {
+///         Ok(MyKernelModule)
+///     }
+/// }
+///
+/// kernel_module!(
+///     MyKernelModule,
+///     author: b"Rust for Linux Contributors",
+///     description: b"My very own kernel module!",
+///     license: b"GPL"
+/// );
+#[macro_export]
+macro_rules! kernel_module {
+    ($module:ty, $($name:ident : $value:expr),*) => {
+        static mut __MOD: Option<$module> = None;
+
+        // Built-in modules are initialized through an initcall pointer
+        //
+        // TODO: find a proper way to emulate the C macro, including
+        // dealing with `HAVE_ARCH_PREL32_RELOCATIONS`
+        #[cfg(not(module))]
+        #[link_section = ".initcall6.init"]
+        #[used]
+        pub static __initcall: extern "C" fn() -> $crate::c_types::c_int = init_module;
+
+        #[no_mangle]
+        pub extern "C" fn init_module() -> $crate::c_types::c_int {
```

Comment:
For builtin kernel modules, we probably need to name mangle this, so you can have more than one, right?

Reply:
We should pass the module name to the kernel macro (or lift it somehow from the folder name automatically) so that we can put a proper name like we do for any other module.

Reply:
Added a couple comments about this.